pkg/bpf/perf.go

```diff
@@ -56,57 +68,97 @@ struct event_sample {
+//      STORE ->data_head               STORE ->data_tail
+//   }
+
+static inline uint64_t perf_read_head(struct perf_event_mmap_page *up)
+{
+	uint64_t data_head = READ_ONCE_64(up->data_head);
+	rmb();
+	return data_head;
+}
+
+static inline void perf_write_tail(struct perf_event_mmap_page *up,
+				   uint64_t data_tail)
+{
+	mb();
+	up->data_tail = data_tail;
+}
+
+int perf_event_read_init(int page_count, int page_size, void *_page,
+			 void *_state)
 {
-	volatile struct perf_event_mmap_page *header = _header;
+	struct perf_event_mmap_page *up = _page;
 	struct read_state *state = _state;
-	uint64_t data_tail = header->data_tail;
-	uint64_t data_head = *((volatile uint64_t *) &header->data_head);
+	uint64_t data_tail = up->data_tail;
 
-	__sync_synchronize();
-	if (data_head == data_tail)
+	if (perf_read_head(up) == data_tail)
 		return 0;
 
-	state->head = (void *) data_head;
 	state->raw_size = page_count * page_size;
-	state->base  = ((uint8_t *)header) + page_size;
-	state->begin = state->base + data_tail % state->raw_size;
-	state->end   = state->base + data_head % state->raw_size;
+	state->base = ((uint8_t *)up) + page_size;
+	state->last_size = 0;
 
-	return state->begin != state->end;
+	return 1;
 }
 
-int perf_event_read(void *_state, void *buf, void *_msg, void *_sample, void *_lost)
+int perf_event_read(void *_page, void *_state, void *_buf, void *_msg,
+		    void *_sample, void *_lost, int page_size)
 {
-	void **msg = (void **) _msg;
+	struct perf_event_mmap_page *up = _page;
+	struct read_state *state = _state;
 	void **sample = (void **) _sample;
 	void **lost = (void **) _lost;
-	struct read_state *state = _state;
-	struct event_sample *e = state->begin;
-
-	if (state->begin == state->end)
+	void **msg = (void **) _msg;
+	void **buf = (void **) _buf;
+	uint64_t e_size, data_tail;
+	struct event_sample *e;
+	void *begin;
+
+	data_tail = up->data_tail + state->last_size;
+	perf_write_tail(up, data_tail);
+	if (perf_read_head(up) == data_tail)
 		return 0;
 
-	if (state->begin + e->header.size > state->base + state->raw_size) {
-		uint64_t len = state->base + state->raw_size - state->begin;
-
-		memcpy(buf, state->begin, len);
-		memcpy((char *) buf + len, state->base, e->header.size - len);
+	e = begin = state->base + (data_tail % state->raw_size);
+	e_size = state->last_size = e->header.size;
+	if (begin + e_size > state->base + state->raw_size) {
+		uint64_t len = state->base + state->raw_size - begin;
+		uint64_t len_first, len_secnd;
+		void *ptr = *buf;
+		int trunc = 0;
+
+		if (e_size > page_size) {
+			ptr = realloc(*buf, e_size);
+			if (!ptr) {
+				ptr = *buf;
+				trunc = 1;
+			} else {
+				*buf = ptr;
+			}
+		}
 
-		e = buf;
-		state->begin = state->base + e->header.size - len;
-	} else if (state->begin + e->header.size == state->base + state->raw_size) {
-		state->begin = state->base;
-	} else {
-		state->begin += e->header.size;
+		len_first = !trunc ? len : (len <= page_size ? len : page_size);
```

Comment:
Purely cosmetic, IMO it's a bit easier to read if the reader isn't forced to handle a "if not xxx {...} else (not not xxx) {...}":

```
len_first = trunc ? (len <= page_size ? len : page_size) : len
```

(Similarly for the case below)